design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,263 @@
+for specific workloads. This functionality will help reduce time to provision a host, because if the host is not found
+matching to expected hardware configuration it means that given host is not capable to run such workloads. If host
+matches to expected configuration, then host is available for use from node pool which can be flagged through
+inventory/label.
+  
+For validating the expected hardware configuration, there will be corresponding code in BMO.
+Any provisioning of nodes will be out of scope for this effort. 
+
+### Implementation Details/Notes/Constraints
+
+Link for Existing Metal3 Specs
+Please refer metal3 spec for bare-metal:
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+1. Add schema for ‘ExpectedHardwareConfiguration’ in  `metal3.io_baremetalhosts_crd.yaml` under ‘Spec’ for BaremetalHost.
+
+    ```yaml
+    ExpectedHardwareConfiguration:
+        default:
+        properties:
+           cpu:
+               properties:
+                   count:
+                       type : Integer
+           type: Object
+           disk:
+               properties:
+                   sizeBytesGB:
+                       type: Integer
+                   numberOfDisks:
+                       type: Integer
+           type: Object
+           nics:
+               properties:
+                  numberofNICS:
+                     type: Integer
+           type: Object
+           ram:
+               properties:
+                   sizeBytesGB:
+                       type: Integer
+           type: Object
+           systemVendor:
+              properties:
+                  name:
+                      type: String
+           type: Object
+           Firmware:
+              Properties:
+                  Version:
+                      RAID:
+                          type: String
+                      BasebandManagement:
+                          type: String
+                      BIOS: 
+                          type: String
+                     IDRAC: 
+                          type: String
+            type: Object
+        required:
+        - cpu
+        - nics
+        - ram
+        - disk
+        - systemVendor
+        - firmware
+    ```
+
+2. Define struct in baremetalhost_types.go to store values for ‘HardwareExpectedConfiguration’ after extracting it
+from `bmhosts_crs.yaml`.
+
+    ```yaml
+    type ExpectedHardwareConfiguration struct {
+     ExpectedCpu           ExpectedCpu          `json:"expectedCpu"`
+     ExpectedDisk          ExpectedDisk         `json:"expectedDisk"`
+     ExpectedNics          ExpectedNics         `json:"expectedNics"`
+     ExpectedRam           int                  `json:"expectedRam"`
+     ExpectedSystemVendor  ExpectedSystemVendor `json:"expectedsystemVendor"`
+    }
+    
+    type ExpectedCpu struct {
+     Count  int `json:"count"`
+    }
+    
+    type ExpectedDisk struct {
```

Comment:
This seems to assume that all of the storage devices are the same size. Is that a safe assumption?

Reply:
Here our thinking is to add minumum disk requirement. Not all disks should have same sizes because root disks sizes are different from data disks. I will add two sizes for disks 1 for root disks and 2 for remaining data disks that way we can set required configuration for disks

Reply:
Yes, if provided CPU in profiles is 24 and if host has 48 CPU, then condition here is CPU >= 24, so that's the baseline criteria for this match.

Reply:
It checks for minimums

Reply:
"Expected" could mean "exact". Let's name the fields to make it clear they are minimums. "MinimumCPUCount" would be an integer with the count. (Maybe we also need MinimumCPUSpeed?) MinimumStorageGB, MinimumStorageDeviceCount, MinimumNICCount, etc.

Reply:
On the other hand, @zaneb pointed out that if these values are always minimums we're likely to have a lot of overlapping classifications. A host that has > 100GB of storage also has > 10GB of storage, for example.

Would it be useful to have that host labeled twice?

How hard would it be to express these rules in a way that includes both min and max?

```
- property: storage
  min: 10
  max: 100
- property: nic
  count: 4
- property: ram
  min: 32
```

Reply:
We can accomodate both parameters in yaml and implement the classification and comparison logic. It's not that hard for us to implement it but it cassifies host with both min and max values. 
I am not able to think scenario for max values ?

Reply:
> "Expected" could mean "exact". Let's name the fields to make it clear they are minimums. "MinimumCPUCount" would be an integer with the count. (Maybe we also need MinimumCPUSpeed?) MinimumStorageGB, MinimumStorageDeviceCount, MinimumNICCount, etc.

Agree, looks good to me.

Reply:
> We can accomodate both parameters in yaml and implement the classification and comparison logic. It's not that hard for us to implement it but it cassifies host with both min and max values.
> Max and MIin value scenario will hit where one profile needs min and max disk, cpu and ram etc, and this kind of request will certainly come in the future.

Right, I think this is going to be important immediately. If I want to classify my hosts based on their size, then I might say hosts with at least 8GB RAM are one type, 16GB RAM another, etc. If I can't express that as a range with both min and max, then every host that has 16GB will also be labeled as being of the type of host that has 8GB RAM, which will make it impossible to distinguish them and place them into separate pools.

Maybe we can build on the matchExpression format used in label selectors? https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements

Reply:
It doesn't make it impossible, it just means that when you select Hosts for pools using labels you'll need to select for atLeast8GB and not atLeast16GB instead of just selecting for has8to15GB.
It's clear that the ergonomics of this are not good. It's less clear that the ergonomics of any alternatives are better.
Even then it's only true when you are using an autoscaler thing to make sure your hardware is 100% utilised. There may in fact be times when you just want a Host with at least 8GB of RAM, and if none with exactly 8GB are available but some with 16GB are then you just want to pick one of those. Although perhaps that's getting into the territory of a fully-fledged scheduler that might not make use of these labels.

Reply:
> > We can accomodate both parameters in yaml and implement the classification and comparison logic. It's not that hard for us to implement it but it cassifies host with both min and max values.
> > Max and MIin value scenario will hit where one profile needs min and max disk, cpu and ram etc, and this kind of request will certainly come in the future.
> 
> Right, I think this is going to be important immediately. If I want to classify my hosts based on their size, then I might say hosts with at least 8GB RAM are one type, 16GB RAM another, etc. If I can't express that as a range with both min and max, then every host that has 16GB will also be labeled as being of the type of host that has 8GB RAM, which will make it impossible to distinguish them and place them into separate pools.
> 
> Maybe we can build on the matchExpression format used in label selectors? https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements

This looks fine, as it will be scalable also for future requirements. 
I think we should finalize this approach if all aggree with this.

Reply:
> It doesn't make it impossible, it just means that when you select Hosts for pools using labels you'll need to select for atLeast8GB and not atLeast16GB instead of just selecting for has8to15GB.
> It's clear that the ergonomics of this are not good. It's less clear that the ergonomics of any alternatives are better.
> Even then it's only true when you are using an autoscaler thing to make sure your hardware is 100% utilised. There may in fact be times when you just want a Host with at least 8GB of RAM, and if none with exactly 8GB are available but some with 16GB are then you just want to pick one of those. Although perhaps that's getting into the territory of a fully-fledged scheduler that might not make use of these labels.

Understood but there are many factors like this, I think to start with we should limit to min and max, we can add more of such rules in the future.

Reply:
I will talk to you both on slack today to finilize the approach for this classifier and what should go in into first phase and what should be planned for future requirements. I think this will help us defining further future requirements also.